File: TREES/BST.cpp
#include<bits/stdc++.h>
using namespace std;
class BST{
    public:
        int data;
        BST *left, *right;
        BST(int value){
            this->data = value;
            this->left = this->right = NULL;
        }
};
BST* insert(BST* root, int value){
    if(!root){
        BST * node = new BST(value);
        return node;
    }
    if(value < root->data){
        root->left = insert(root->left,value);
    }
    else{
        root->right = insert(root->right,value);
    }
    return root;
}
void preorder(BST*root){
    if(root){
        cout<< root->data<<" ";
        preorder(root->left);
        preorder(root->right);
    }
}
void inorder(BST* root){
    if(!root) return;
    inorder(root->left);
    cout<<root->data<<" ";
    inorder(root->right);
}
bool search(BST* root,int val){
    if(!root) return 0;
    if(val==root->data) return 1;
    if(val<root->data){
        return search(root->left,val);
    }
    else{
        return search(root->right,val);
    }

}
BST* remove(BST* root, int x){
    
}
int main(){
    int arr[] = {3,4,7,1,6,8};
    BST * root = NULL;
    for(int i=0; i<6; i++){
        root = insert(root,arr[i]);
    }

    inorder(root);
    cout<<endl;
    
    return 0;
}

File: TREES/BST.py
class Node:
    def __init__(self,item=None, left=None,right=None):
        self.item = item
        self.left = left
        self.right = right

class BST:
    def __init__(self):
        self.root = None
    def rinsert(self,root,data):
        if root is None:
            return Node(data)
        if data < root.item:
            root.left=self.rinsert(root.left,data)
        elif data > root.item:
            root.right=self.rinsert(root.right,data)
        return root
    def insert(self,data):
        self.root = self.rinsert(self.root, data)
    def rsearch(self,root,data):
        if root is None or root.item == data:
            return root
        if data<root.item:
            return self.rsearch(root.left,data)
        else:
            return self.rsearch(root.right,data)

    def search(self,data):
        return self.rsearch(self.root, data)
    def inorder(self):
        result = []
        self.rinorder(self.root, result)
        return result
    def rinorder(self,root,result):
        if root:
            self.rinorder(root.left,result)
            result.append(root.item)
            self.rinorder(root.right,result)
    def preorder(self):
        result=[]
        self.rpreorder(self.root,result)
        return result
    def rpreorder(self,root,result):
        if root:
            result.append(root.item)
            self.rpreorder(root.left, result)
            self.rpreorder(root.right, result)
    def postorder(self):
        result=[]
        self.rpostorder(self.root,result)
        return result
    def rpostorder(self,root,result):
        if root:
            self.rpostorder(root.left, result)
            self.rpostorder(root.right, result)
            result.append(root.item)

bst = BST()
bst.insert(50)
bst.insert(10)
bst.insert(80)
bst.insert(3)
# item = bst.search(10)
# print(item.item)

for x in bst.inorder():
    print(x,end=" ")


    
    